spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/SqsMessageProducer.java

```diff
@@ -0,0 +1,160 @@
+
+	private volatile boolean running;
+
+	public SqsMessageProducer(String logicalEndpointName, QueueAttributes queueAttributes, SqsAsyncClient sqsClient) {
+		this.logicalEndpointName = logicalEndpointName;
+		this.queueUrl = queueAttributes.getDestinationUrl();
+		this.queueAttributes = queueAttributes;
+		this.sqsAsyncClient = sqsClient;
+	}
+
+	@Override
+	public CompletableFuture<Collection<Message<String>>> produce(int numberOfMessages, Duration timeout) {
+		logger.trace("Polling for messages at {}", this.queueUrl);
+		return sqsAsyncClient
+				.receiveMessage(req -> req.queueUrl(this.queueUrl).maxNumberOfMessages(numberOfMessages)
+						.waitTimeSeconds((int) timeout.getSeconds()))
+				.thenApply(ReceiveMessageResponse::messages).thenApply(this::getMessagesForExecution)
+				.exceptionally(handleException());
+	}
+
+	protected Function<Throwable, Collection<Message<String>>> handleException() {
+		return t -> {
+			logger.error("Error retrieving messages from SQS", t);
+			return Collections.emptyList();
+		};
+	}
+
+	private Collection<Message<String>> getMessagesForExecution(
+			List<software.amazon.awssdk.services.sqs.model.Message> messages) {
+		logger.trace("Poll returned {} messages", messages.size());
+		return messages.stream().map(this::getMessageForExecution).collect(Collectors.toList());
+	}
+
+	protected org.springframework.messaging.Message<String> getMessageForExecution(
+			final software.amazon.awssdk.services.sqs.model.Message message) {
+		HashMap<String, Object> additionalHeaders = new HashMap<>();
+		additionalHeaders.put(SqsMessageHeaders.SQS_LOGICAL_RESOURCE_ID, this.logicalEndpointName);
+		additionalHeaders.put(SqsMessageHeaders.RECEIVED_AT, Instant.now());
+		additionalHeaders.put(SqsMessageHeaders.QUEUE_VISIBILITY, this.queueAttributes.getVisibilityTimeout());
+		additionalHeaders.put(SqsMessageHeaders.VISIBILITY,
+				new QueueMessageVisibility(this.sqsAsyncClient, this.queueUrl, message.receiptHandle()));
+		return createMessage(message, additionalHeaders);
+	}
+
+	protected org.springframework.messaging.Message<String> createMessage(
+			software.amazon.awssdk.services.sqs.model.Message message, Map<String, Object> additionalHeaders) {
+
+		HashMap<String, Object> messageHeaders = new HashMap<>();
+		messageHeaders.put(SqsMessageHeaders.MESSAGE_ID_MESSAGE_ATTRIBUTE_NAME, message.messageId());
+		messageHeaders.put(SqsMessageHeaders.RECEIPT_HANDLE_MESSAGE_ATTRIBUTE_NAME, message.receiptHandle());
+		messageHeaders.put(SqsMessageHeaders.SOURCE_DATA_HEADER, message);
+		messageHeaders.put(MessageHeaders.ACKNOWLEDGMENT_HEADER,
+				new SqsAcknowledge(this.sqsAsyncClient, this.queueUrl, message.receiptHandle()));
+		messageHeaders.putAll(additionalHeaders);
+		messageHeaders.putAll(getAttributesAsMessageHeaders(message));
+		messageHeaders.putAll(getMessageAttributesAsMessageHeaders(message));
+		return new GenericMessage<>(message.body(), new SqsMessageHeaders(messageHeaders));
```

Comment:
```suggestion
		return new GenericMessage<>(message.body(), new SqsMessageHeaders(Collections.unmodifiableMap(messageHeaders)));
```

Reply:
I most certainly accepted and submitted most of these suggestions, but they don't seem to appear here. I'll give it a look.